frontends/web/src/components/language/language.test.tsx

```diff
@@ -15,46 +15,75 @@
  */
 
 import 'jest';
-import { mount } from 'enzyme';
-jest.mock('../../i18n/i18n');
-import i18n from '../../i18n/i18n';
 import { LanguageSwitch } from './language';
-import { act } from 'react-dom/test-utils';
+import { render, fireEvent } from '@testing-library/react';
+
+import { useTranslation } from 'react-i18next';
+
+jest.mock('react-i18next');
+const useTranslationSpy = useTranslation;
 
 describe('components/language/language', () => {
-  describe('selectedIndex', () => {
-    const supportedLangs = [
-      { code: 'en', display: 'EN' },
-      { code: 'en-US', display: 'EN' },
-      { code: 'pt', display: 'PT' },
-    ];
 
-    supportedLangs.forEach((lang, idx) => {
-      it(`returns exact match (${lang.code})`, async () => {
-        await i18n.changeLanguage(lang.code);
 
-        let ctx: any;
-        act(() => {
-          /* @ts-ignore */
-          ctx = mount(<LanguageSwitch i18n={i18n} languages={supportedLangs} />);
+  const supportedLangs = [
+    { code: 'en-US', display: 'English' },
+    { code: 'pt', display: 'Portugues' },
+    { code: 'ms', display: 'Bahasa Melayu' },
+    { code: 'de', display: 'Deutsch' },
+  ];
+
+  /**
+   * renderSwitchAndOpenDialog is a util function used to render
+   * the language switch and open the language dialog. This returns
+   * `RenderResult` from the render function of `@testing-library/react`.
+   */
+  function renderSwitchAndOpenDialog() {
+    //@ts-ignore
+    const rendered = render(<LanguageSwitch languages={supportedLangs} />);
+    const btn = rendered.getByTitle('Select Language');
+    fireEvent.click(btn);
+    return rendered;
+  }
+
+  describe('selectedIndex', () => {
+    supportedLangs.forEach((lang, _) => {
+      it(`returns exact match (${lang.code})`, async () => {
+        (useTranslationSpy as jest.Mock).mockReturnValue({
+          t: jest.fn(),
+          i18n: {
+            language: lang.code
+          },
         });
 
-        expect(ctx.childAt(0).state('selectedIndex')).toEqual(idx);
+        const { getByTestId } = renderSwitchAndOpenDialog();
+        const selectedLang = getByTestId(`language-selection-${lang.code}`);
+        expect(selectedLang.classList.contains('selected')).toBe(true);
       });
     });
 
     it('matches main language tag', async () => {
-      await i18n.changeLanguage('pt_BR');
-      /* @ts-ignore */
-      const ctx = mount(<LanguageSwitch i18n={i18n} languages={supportedLangs} />);
-      expect(ctx.childAt(0).state('selectedIndex')).toEqual(2); // 'pt'
+      (useTranslationSpy as jest.Mock).mockReturnValue({
+        t: jest.fn(),
+        i18n: {
+          language: 'de'
+        },
+      });
+      const { getByTestId } = renderSwitchAndOpenDialog();
+      const selectedLang = getByTestId('language-selection-de');
+      expect(selectedLang.classList.contains('selected')).toBe(true);
```

Comment:
Instead of testing the value of the useState (state) internally like we used to, we should test what that state change reflects, which is what the users see. These lines (and similar lines below) do exactly that.


[Read more about why we should avoid testing internal states.](https://stackoverflow.com/a/60109830)